File: package/go/tests/count_word_frequencies_test.go
package tests

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"program-knock/src"
)

func TestCountWordFrequencies(t *testing.T) {
	t.Run("Basic", func(t *testing.T) {
		text := "apple banana apple orange banana apple"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"apple": 3, "banana": 2, "orange": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("SingleWord", func(t *testing.T) {
		text := "python"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"python": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("TiedFrequencies", func(t *testing.T) {
		text := "dog cat bird cat dog bird"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"bird": 2, "cat": 2, "dog": 2}
		assert.Equal(t, expected, result)
	})

	t.Run("EmptyString", func(t *testing.T) {
		text := ""
		result := src.CountWordFrequencies(text)
		expected := map[string]int{}
		assert.Equal(t, expected, result)
	})

	t.Run("RepeatedSingleWord", func(t *testing.T) {
		// 100 times "test"
		words := make([]string, 100)
		for i := range words {
			words[i] = "test"
		}
		text := strings.Join(words, " ")
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"test": 100}
		assert.Equal(t, expected, result)
	})

	t.Run("WordsOfVariousLengths", func(t *testing.T) {
		text := "a bb ccc a bb"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"a": 2, "bb": 2, "ccc": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("AlphabeticalOrderForSameFrequency", func(t *testing.T) {
		text := "cat dog cat dog fish bird fish bird"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"bird": 2, "cat": 2, "dog": 2, "fish": 2}
		assert.Equal(t, expected, result)
	})

	t.Run("VaryingFrequencies", func(t *testing.T) {
		text := "apple apple banana apple banana orange lemon"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"apple": 3, "banana": 2, "lemon": 1, "orange": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("OneLetterWords", func(t *testing.T) {
		text := "a b c a b a"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"a": 3, "b": 2, "c": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("ManyWordsCase", func(t *testing.T) {
		// Build test text: apple*12, banana*8, cherry*7
		var parts []string
		for i := 0; i < 10; i++ {
			parts = append(parts, "apple")
		}
		for i := 0; i < 5; i++ {
			parts = append(parts, "banana")
		}
		for i := 0; i < 5; i++ {
			parts = append(parts, "cherry")
		}
		for i := 0; i < 3; i++ {
			parts = append(parts, "banana")
		}
		for i := 0; i < 2; i++ {
			parts = append(parts, "apple")
		}
		for i := 0; i < 2; i++ {
			parts = append(parts, "cherry")
		}
		text := strings.Join(parts, " ")

		result := src.CountWordFrequencies(text)
		expected := map[string]int{"apple": 12, "banana": 8, "cherry": 7}
		assert.Equal(t, expected, result)
	})

	t.Run("ExtraSpacesHandling", func(t *testing.T) {
		text := "  apple   banana  apple  "
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"apple": 2, "banana": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("CaseSensitive", func(t *testing.T) {
		text := "Apple apple APPLE"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"Apple": 1, "apple": 1, "APPLE": 1}
		assert.Equal(t, expected, result)
	})

	t.Run("SpecialCharacters", func(t *testing.T) {
		text := "hello! world? hello!"
		result := src.CountWordFrequencies(text)
		expected := map[string]int{"hello!": 2, "world?": 1}
		assert.Equal(t, expected, result)
	})
}

File: package/java/src/test/java/com/programknock/CountWordFrequenciesTest.java
package com.programknock;

import org.junit.jupiter.api.Test;
import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class CountWordFrequenciesTest {

    @Test
    void testBasic() {
        String text = "apple banana apple orange banana apple";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("apple", 3);
        expected.put("banana", 2);
        expected.put("orange", 1);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testSingleWord() {
        String text = "python";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("python", 1);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testTiedFrequencies() {
        String text = "dog cat bird cat dog bird";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("bird", 2);
        expected.put("cat", 2);
        expected.put("dog", 2);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testEmptyString() {
        String text = "";
        Map<String, Integer> expected = new LinkedHashMap<>();

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testRepeatedSingleWord() {
        List<String> words = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            words.add("test");
        }
        String text = String.join(" ", words);

        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("test", 100);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testWordsOfVariousLengths() {
        String text = "a bb ccc a bb";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("a", 2);
        expected.put("bb", 2);
        expected.put("ccc", 1);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testAlphabeticalOrderForSameFrequency() {
        String text = "cat dog cat dog fish bird fish bird";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("bird", 2);
        expected.put("cat", 2);
        expected.put("dog", 2);
        expected.put("fish", 2);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testVaryingFrequencies() {
        String text = "apple apple banana apple banana orange lemon";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("apple", 3);
        expected.put("banana", 2);
        expected.put("lemon", 1);
        expected.put("orange", 1);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testOneLetterWords() {
        String text = "a b c a b a";
        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("a", 3);
        expected.put("b", 2);
        expected.put("c", 1);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testManyWordsCase() {
        List<String> words = new ArrayList<>();
        for (int i = 0; i < 10; i++) words.add("apple");
        for (int i = 0; i < 5; i++) words.add("banana");
        for (int i = 0; i < 5; i++) words.add("cherry");
        for (int i = 0; i < 3; i++) words.add("banana");
        for (int i = 0; i < 2; i++) words.add("apple");
        for (int i = 0; i < 2; i++) words.add("cherry");

        String text = String.join(" ", words);

        Map<String, Integer> expected = new LinkedHashMap<>();
        expected.put("apple", 12);
        expected.put("banana", 8);
        expected.put("cherry", 7);

        assertEquals(expected, CountWordFrequencies.countWordFrequencies(text));
    }

    @Test
    void testNullInput() {
        assertEquals(new LinkedHashMap<>(), CountWordFrequencies.countWordFrequencies(null));
    }

    @Test
    void testWhitespaceOnly() {
        assertEquals(new LinkedHashMap<>(), CountWordFrequencies.countWordFrequencies("   "));
    }
}
